[BREAKING] Add column indexing using strings

bkamins opened the pull request:
Fixes https://github.com/JuliaData/DataFrames.jl/issues/1926.

**UPDATED**

The list of changes in this PR is:

User visible:
* allow using an `AbstractString` **everywhere** where `Symbol` was accepted as a column indicator (i.e. this extends beyond just `getindex` and `getproperty`, every function was reviewed and made consistent)
* docstrings were updated to reflect this change + some unrelated docs fixes were added where spotted
* `names` function consistently returns vector of strings for all types that return column names
* `propertynames` function consistently returns vector of `Symbol`s for all types that return column names
* `keys` function consistently returns vector of `Symbol`s for all types that return column names as keys
* depend on DataAPI.jl v1.2
* `rename!` passes `String` to a function if API with function is used
* consistently define `hasproperty` for types that have custom `getproperty`
* add `cols` to `names` function for types where it makes sense consistently
* `push!` accepts dicts with string keys
* `indexing with an arbitrary eltype vector now throws better error messages (informtive `ArgumentError` is thrown in more cases rather than `MethodError` in the past)

Internal:
* add definitions of `SymbolOrString`, `ColumnIndex` (updated), and `MultiColumnIndex` unions and use them consistently across whole codebase
* remove `keys` from `AbstractIndex` as it was never used
* correct formatting of code in many places (mainly docstrings, function signatures, and adding missing `return` statements in long functions as many of them were updated anyway, but also in unrelated places)
* fix a rule that `_names` returns vector of `Symbol`s without copying
* in some places switch from `names` to `_names` internally to avoid unnecessary allocations
* in some places use `groupcols` where this function would be an appropriate to call; similarly with `names(df, cols)` where it made sense
* remove `without` function that was not needed anymore

bkamins commented:
OK - if anyone is interested - currently I have updated the basic tests and documentation. The PR should be good to have a first look at and comment.

Now I will be working on testing corner cases of the new functionality.

bkamins commented:
Lessons (re)learned so far:
* I already do not remember what works on 1.0 and we want to keep backward compatibility (the language is moving forward fast)
* method ambiguities with functions from Base that have very loose signatures is a pain
* I expect that some bugs introduced will be caught only after the release and testing in practice (the PR really touches everything we have in DataFrames.jl)

bkamins commented:
`Between` from DataAPI.jl does not allow string indexing. I am extending it in DataFrames.jl which is a slight type piracy.

quinnj commented:
> `Between` from DataAPI.jl does not allow string indexing. I am extending it in DataFrames.jl which is a slight type piracy.

Happy to add it to DataAPI.jl quickly if we want.

bkamins commented:
> Happy to add it to DataAPI.jl quickly if we want.

Thank you.

@nalimilan - what do you think? The definition is:

```
DataAPI.Between(x::AbstractString, y::AbstractString) = Between(Symbol(x), Symbol(y))
DataAPI.Between(x::Union{Int, Symbol}, y::AbstractString) = Between(x, Symbol(y))
DataAPI.Between(x::AbstractString, y::Union{Int, Symbol}) = Between(Symbol(x), y)
```
(in DataAPI.jl it would be simpler)

so the idea is is to allow string as input but internally we store it as `Symbol`, so other packages that require `Symbol` for column indexing will work.

bkamins commented:
The PR is ready for reviews / beta testing.

bkamins commented:
If https://github.com/JuliaData/DataAPI.jl/pull/17 is merged then this PR should be updated to remove `Between` cases for `AbstractString`.

nalimilan commented on test/data.jl:
```diff
@@ -375,6 +397,10 @@ end
     filter!(AsTable(:x) => testfun, df)
     @test df == DataFrame(x=[3, 2], y=["b", "a"])
 
+    @test filter(AsTable("x") => testfun, df) == DataFrame(x=[3, 2], y=["b", "a"])
```
Filtering twice isn't a very strict test... :-p

bkamins commented:
> How confident are you that tests cover (almost) everything?

I went through each change I made and manually added tests.

> Does coverage confirm that all new methods are tested?

Where it works it shows that coverage is OK, but it is not 100% full check as we have https://github.com/JuliaData/DataFrames.jl/issues/2180 problem.

bkamins commented on src/abstractdataframe/join.jl:
```diff
@@ -346,9 +353,9 @@ function _join(df1::AbstractDataFrame, df2::AbstractDataFrame;
 
 """
-    innerjoin(df1, df2; on = Symbol[], makeunique = false,
+    innerjoin(df1, df2; on, makeunique = false,
```
setting `on` as `Symbol[]` is a workaround to throw an informative error. `on` must be passed, and writing that it defaults to `Symbol[]` is confusing as this is not an accepted value.

bkamins commented on src/dataframe/dataframe.jl:
```diff
@@ -153,9 +162,27 @@ function DataFrame(pairs::Pair{Symbol,<:Any}...; makeunique::Bool=false,
                    copycols::Bool=true)::DataFrame
     colnames = [Symbol(k) for (k,v) in pairs]
     columns = Any[v for (k,v) in pairs]
```
No - as `v` can be a scalar and pseudo-broadcasting happens in an inner constructor.

bkamins commented on src/other/broadcasting.jl:
```diff
@@ -98,20 +100,20 @@ Base.maybeview(df::AbstractDataFrame, rows, cols) = view(df, rows, cols)
         throw(ArgumentError("creating new columns using an integer index is disallowed"))
     end
-    LazyNewColDataFrame(df, cols)
+    return LazyNewColDataFrame(df, cols isa AbstractString ? Symbol(cols) : cols)
```
This is incorrect as in this method `cols` can be also a number at this point (it was not possible in line 106 so I have accepted your suggestion there).

bkamins commented on src/other/index.jl:
```diff
@@ -92,13 +101,16 @@ function rename!(x::Index, nms::AbstractVector{Pair{Symbol, Symbol}})
 
+# we do not define keys on purpose;
+# use names to get keys as strings with copying
+# or _names to get keys as Symbols without copying
```
`keys` has different uses and different return values and `_names` is consistent (I will add a section to the manual describing this).

Based on the discussion all functions will return vectors only (as we are type unstable anyway).

bkamins commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -65,22 +65,33 @@ abstract type AbstractDataFrame end
-    Return a `Vector{Symbol}` of names of columns contained in `df`.
+    Return a `Vector{String}` of names of columns contained in `df`.
+
+    NOTE: Internally column names are stored as `Symbol`s. `names` returns them
+    as `String`s for convenience of further processing. If you want to get a tuple
+    of `Symbol`s representing column names use `propertynames`.
```
OK - changed.

bkamins commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -6,9 +6,7 @@ for working with tabular data.
-An AbstractDataFrame is a two-dimensional table with Symbols for
-column names. An AbstractDataFrame is also similar to an Associative
-type in that it allows indexing by a key (the columns).
+An `AbstractDataFrame` is a two-dimensional table with `Symbol`s for column names.
```
OK

bkamins commented on src/abstractdataframe/iteration.jl:
```diff
@@ -158,16 +165,23 @@ end
     keys(dfc::DataFrameColumns)
 
-Get a vector of column names of `dfc`.
+Get a tuple of column names of `dfc`.
```
fixed

bkamins commented on src/other/index.jl:
```diff
@@ -225,13 +243,18 @@ end
-                        "Only Integer or Symbol values allowed when indexing by vector"))
+    all(x -> x isa AbstractString, idxs) && return getindex(x, Symbol.(idxs))
+
+    if idxs[1] isa AbstractString
```
OK

bkamins commented:
I have updated the codes following the review. What is left for tomorrow is updating the docstrings to be consistent and changing tests not to use `Symbol.(` conversion.

bkamins commented on test/select.jl:
```diff
@@ -253,48 +253,48 @@ end
     d = copy(df, copycols=false)
     select!(d, [:a, :e, :c])
-    @test names(d) == [:a, :e, :c]
+    @test Symbol.(names(d)) == [:a, :e, :c]
```
done

bkamins commented on test/indexing.jl:
```diff
@@ -507,18 +536,26 @@ end
+    @test view(dfr, ["a", "b"]) isa DataFrameRow
+    @test copy(view(dfr, ["a", "b"])) == (a=1, b=4)
     @test view(dfr, r"[ab]") isa DataFrameRow
-    @test copy(dfr[r"[ab]"]) == (a=1, b=4)
```
yes - it is tested in the `@testset` above - it was my mistake some time ago that I copy pasted `getindex` and did not change it to `view` here.

bkamins commented on test/deprecated.jl:
```diff
@@ -407,7 +407,7 @@ end
         df[3] = [1,2,3]
         df[4] = [1,2,3]
-        @test names(df) == [:x3, :x3_1, :x3_2, :x4]
+        @test Symbol.(names(df)) == [:x3, :x3_1, :x3_2, :x4]
```
fixed

bkamins commented on test/constructors.jl:
```diff
@@ -124,28 +177,28 @@ end
     df = DataFrame(x=x, y=y)
     @test size(df) == (3, 2)
-    @test names(df) == [:x, :y]
+    @test Symbol.(names(df)) == [:x, :y]
```
I change `propertynames` everywhere as it is simpler.

bkamins commented on test/constructors.jl:
```diff
@@ -34,39 +34,56 @@ const ≅ = isequal
                           x2 = Union{Int, Missing}[1.0, 1.0, 1.0])
     @test df == DataFrame!(x1 = Union{Int, Missing}[0.0, 0.0, 0.0],
                            x2 = Union{Int, Missing}[1.0, 1.0, 1.0])
-    @test df == DataFrame([:x1=>Union{Int, Missing}[0.0, 0.0, 0.0],
```
my mistake - I thought it was a duplicate test, but it is not. reverted.

bkamins commented on test/constructors.jl:
```diff
@@ -34,39 +34,56 @@ const ≅ = isequal
     @test df == DataFrame((CategoricalVector{Union{Float64, Missing}}(zeros(3)),
                            CategoricalVector{Union{Float64, Missing}}(ones(3))))
     @test df == DataFrame((CategoricalVector{Union{Float64, Missing}}(zeros(3)),
-                           CategoricalVector{Union{Float64, Missing}}(ones(3))), (:x1, :x2))
```
same here - it is allowed. I will revert it.

bkamins commented on test/cat.jl:
```diff
@@ -129,91 +129,91 @@ end
 
     df3 = hcat(df1, df2)
-    @test names(df3) == [:a, :b]
+    @test Symbol.(names(df3)) == [:a, :b]
```
LHS changed to `propertynames` in all such cases

bkamins commented on test/data.jl:
```diff
@@ -375,6 +397,10 @@ end
     filter!(AsTable(:x) => testfun, df)
     @test df == DataFrame(x=[3, 2], y=["b", "a"])
 
+    @test filter(AsTable("x") => testfun, df) == DataFrame(x=[3, 2], y=["b", "a"])
```
it was a test of dispatch, but I changed it :).

bkamins commented on test/dataframe.jl:
```diff
@@ -27,12 +27,12 @@ const ≇ = !isequal
         while true
             rename!(df, p)
-            @test sort(names(df)) == n
+            @test sort(Symbol.(names(df))) == n
```
done

bkamins commented on test/data.jl:
```diff
@@ -16,9 +16,9 @@ const ≅ = isequal
-    @test names(df6) == [:A, :B, :C]
-    @test names(df2) == [:x1, :x2]
-    @test names(df7) == [:x, :y]
+    @test Symbol.(names(df6)) == [:A, :B, :C]
```
changed to `propertynames`

bkamins commented on test/conversions.jl:
```diff
@@ -100,23 +100,23 @@ const ≅ = isequal
     df = convert(DataFrame, di)
     @test isa(df, DataFrame)
-    @test names(df) == [Symbol(x) for x in sort(collect(keys(di)))]
+    @test Symbol.(names(df)) == [Symbol(x) for x in sort(collect(keys(di)))]
```
done

bkamins commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -1094,7 +1125,7 @@ See also [`unique`](@ref) and [`unique!`](@ref).
 # Arguments
 - `df` : the AbstractDataFrame
-- `cols` : a column indicator (Symbol, Int, Vector{Symbol}, etc.)
+- `cols` : a column indicator (`Symbol`, `Int`, `Vector{Symbol}`, etc.)
```
proposed. I will try to update it everywhere where applicable

bkamins commented on src/abstractdataframe/reshape.jl:
```diff
@@ -115,14 +117,8 @@ function _stackview(df::AbstractDataFrame, measure_vars::AbstractVector{Int},
-            renamecols::Function=identity)
-    unstack(df::AbstractDataFrame, colkey::Union{Integer, Symbol},
-            value::Union{Integer, Symbol}; renamecols::Function=identity)
+    unstack(df::AbstractDataFrame, rowkeys, colkey, value; renamecols::Function=identity)
```
fixed

bkamins commented on src/dataframe/dataframe.jl:
```diff
@@ -863,9 +929,7 @@ Base.hcat(df1::DataFrame, df2::AbstractDataFrame, dfn::AbstractDataFrame...;
-    allowmissing!(df::DataFrame, cols::Colon=:)
-    allowmissing!(df::DataFrame, cols::Union{Integer, Symbol})
-    allowmissing!(df::DataFrame, cols::Union{AbstractVector, Regex, Not, Between, All})
+    allowmissing!(df::DataFrame, cols=:)
```
done

bkamins commented on src/dataframe/dataframe.jl:
```diff
@@ -954,22 +1015,16 @@ disallowmissing!(df::DataFrame, cols::Colon=:; error::Bool=true) =
 
-Change columns selected by `cname` or `cnames` in data frame `df`
-to `CategoricalVector`.
+Change columns selected by `cnames` in data frame `df` to `CategoricalVector`.
```
done

bkamins commented on src/dataframe/dataframe.jl:
```diff
@@ -901,10 +965,7 @@ allowmissing!(df::DataFrame, cols::Colon=:) =
-    disallowmissing!(df::DataFrame, cols::Union{Integer, Symbol}; error::Bool=true)
-    disallowmissing!(df::DataFrame, cols::Union{AbstractVector, Regex, Not, Between, All};
-                     error::Bool=true)
+    disallowmissing!(df::DataFrame, cols=:; error::Bool=true)
```
done

bkamins commented:
All should be fixed now. In particular I have created one common pair of strings for single and multiple column indexing that are reused everywhere where applicable (hopefully).

nalimilan commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -65,22 +66,29 @@ abstract type AbstractDataFrame end
 
-    If a `cols` column selector is passed then restrict returned
-    column names to those matching the selector
+    If a `cols` column selector ($COLUMN_INDICATOR, $COLUMNS_INDICATOR)
```
The concatenation of the two may be a bit weird because of the "or integer" in the middle. Maybe separate them with a ";" (here and elsewhere)?

Also, inserting this long list in the middle of a sentence will be hard to read. Probably better add this at the end of the paragraph in a separate sentence like "`cols` can be any column selector (...)." Same in other places (including argument lists).

bkamins replied:
OK - I will change this

nalimilan commented on test/constructors.jl:
```diff
@@ -34,39 +34,68 @@ const ≅ = isequal
                            CategoricalVector{Union{Float64, Missing}}(ones(3))))
     @test df == DataFrame((CategoricalVector{Union{Float64, Missing}}(zeros(3)),
                            CategoricalVector{Union{Float64, Missing}}(ones(3))), (:x1, :x2))
+    @test df == DataFrame((CategoricalVector{Union{Float64, Missing}}(zeros(3)),
+                           CategoricalVector{Union{Float64, Missing}}(ones(3))), (:x1, :x2))
+    @test df == DataFrame((CategoricalVector{Union{Float64, Missing}}(zeros(3)),
+                           CategoricalVector{Union{Float64, Missing}}(ones(3))), ("x1", "x2"))
```
These are duplicated AFAICT.
```suggestion
```

bkamins replied:
they are not duplicated as they test tuple interface (earlier tests check vector interface)

nalimilan replied:
Both seem to test tuples to me.

bkamins replied:
I have rewritten this part of tests entirely to make them more readable. I hope now it will be more clear (and it is shorter).

nalimilan commented on test/constructors.jl:
```diff
@@ -34,39 +34,68 @@ const ≅ = isequal
                             :x2=>Union{Int, Missing}[1.0, 1.0, 1.0]))
+    @test df == DataFrame(["x1"=>Union{Int, Missing}[0.0, 0.0, 0.0],
+                           "x2"=>Union{Int, Missing}[1.0, 1.0, 1.0]])
+    @test df == DataFrame(["x1"=>Union{Int, Missing}[0.0, 0.0, 0.0],
+                           "x2"=>Union{Int, Missing}[1.0, 1.0, 1.0]])
+    @test df == DataFrame!(("x1"=>Union{Int, Missing}[0.0, 0.0, 0.0],
+                            "x2"=>Union{Int, Missing}[1.0, 1.0, 1.0]))
```
Also here:
```suggestion
```

bkamins replied:
they are not duplicated but I put `!` in a wrong place - I will fix it.

nalimilan commented on test/subdataframe.jl:
```diff
@@ -212,7 +212,7 @@ end
     @test !haskey(DataFrames.index(df2), 0)
     @test_throws ArgumentError haskey(DataFrames.index(df2), true)
-    @test keys(DataFrames.index(df2)) == [:y]
+    @test DataFrames._names(DataFrames.index(df2)) == [:y]
```
```suggestion
    @test names(DataFrames.index(df2)) == ["y"]
    @test DataFrames._names(DataFrames.index(df2)) == [:y]
```

bkamins commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -333,9 +338,17 @@ Return the number of dimensions of a data frame, which is always `2`.
 Base.getproperty(df::AbstractDataFrame, col_ind::Symbol) = df[!, col_ind]
+Base.getproperty(df::AbstractDataFrame, col_ind::AbstractString) = df[!, col_ind]
+
 # Private fields are never exposed since they can conflict with column names
```
I have not made this comment originally. But it means that column name can have the same name as private field of a data frame so `true` is ignored if `private=true`.

nalimilan replied:
Ah I think it was mine. OK, I misunderstood it as referring to the fact that we're making a copy of a private field.

bkamins commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -65,22 +66,29 @@ abstract type AbstractDataFrame end
     names(df::AbstractDataFrame, cols)
 
-    Return a `Vector{Symbol}` of names of columns contained in `df`.
+    Return a freshly allocated `Vector{String}` of names of columns contained in `df`.
```
fixed

bkamins commented:
OK - I am done with round 2 of documentation fixes :)

nalimilan commented on src/dataframe/dataframe.jl:
```diff
@@ -1017,9 +1019,9 @@ disallowmissing!(df::DataFrame, cols::Colon=:; error::Bool=true) =
-in data frame `df` to `CategoricalVector`.
+Change columns selected by `cols` in data frame `df` to `CategoricalVector`.
+
+`cols` can be any column selector ($COLUMN_INDICATOR; $COLUMNS_INDICATOR) or `Type`.
```
```suggestion
`cols` can be any column selector ($COLUMN_INDICATOR; $COLUMNS_INDICATOR) or a `Type`.
```

nalimilan commented on src/groupeddataframe/splitapplycombine.jl:
```diff
@@ -9,10 +9,12 @@ Return a `GroupedDataFrame` representing a view of an `AbstractDataFrame` split
 - `sort` : whether to sort rows according to the values of the grouping columns `cols`
 - `skipmissing` : whether to skip rows with `missing` values in one of the grouping columns `cols`
 
+`cols` can be any column selector ($COLUMN_INDICATOR; $COLUMNS_INDICATOR).
```
```suggestion
- `cols` : data frame columns to group by. Can be any column selector
  ($COLUMN_INDICATOR; $COLUMNS_INDICATOR).
- `sort` : whether to sort rows according to the values of the grouping columns `cols`
- `skipmissing` : whether to skip rows with `missing` values in one of the grouping columns `cols`
```

bkamins commented on test/constructors.jl:
```diff
@@ -16,66 +16,34 @@ const ≅ = isequal
+    vecvec = [CategoricalVector{Union{Float64, Missing}}(zeros(3)),
+              CategoricalVector{Union{Float64, Missing}}(ones(3))]
+
+    df = DataFrame(collect(Any, vecvec), Index([:x1, :x2]))
```
There is no method for `DataFrame(vecvec, DataFrames.Index([:x1, :x2]))` (inner constructor only accepts eltype `Any` or `AbstractVector`)

bkamins commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -664,10 +667,11 @@ end
+(complete cases) in data frame `df`. If `cols` is provided, only missing values in
 the corresponding columns are considered.
 
+`cols` can be any column selector ($COLUMN_INDICATOR; $COLUMNS_INDICATOR).
```
For some reason GitHub disallows committing some suggestions. I will apply them manually.

bkamins commented:
Following your fixes I have reformatted some code to also fix the customary line witdth.

This prompted me to make one change (only internal - the API is unchanged). I defined `MultiColumnIndex` union and use it consistently. It is much cleaner this way I think (as otherwise it is a pain to check if all signatures are correct). Sorry for making this change so late, but I think we should have such an abstraction that complements `ColumnIndex`.

Finally - it will make implementation of https://github.com/JuliaData/DataAPI.jl/issues/16 a minor PR (we will have to change a few lines of code to incorporate it).

nalimilan commented on src/abstractdataframe/sort.jl:
```diff
@@ -322,6 +322,7 @@ See other methods for a description of other keyword arguments.
 """
 function Base.issorted(df::AbstractDataFrame, cols=[];
                        lt=isless, by=identity, rev=false, order=Forward)
+    # exclude AbstractVector as in that case cols can contain ordered(...) clauses
```
Isn't that `order`?

bkamins replied:
right - fixed

nalimilan commented on src/abstractdataframe/sort.jl:
```diff
@@ -322,6 +322,7 @@ See other methods for a description of other keyword arguments.
 """
 function Base.issorted(df::AbstractDataFrame, cols=[];
                        lt=isless, by=identity, rev=false, order=Forward)
+    # exclude AbstractVector as in that case cols can contain ordered(...) clauses
```
Isn't that `order`?

nalimilan replied:
Isn't that `order`?

bkamins commented on src/other/index.jl:
```diff
@@ -10,12 +10,13 @@ end
 Base.summary(io::IO, idx::AbstractIndex) = print(io, summary(idx))
 
 const ColumnIndex = Union{Signed, Unsigned, Symbol, AbstractString}
+const MultiColumnIndex = Union{AbstractVector, Regex, Not, Between, All, Colon}
+const MULTICOLUMNINDEX_TUPLE = (:AbstractVector, :Regex, :Not, :Between, :All, :Colon)
```
this is tempting but it does not work (and I could not come up with a good substitute without metaprogramming), so I would leave it as is.

bkamins commented on src/other/index.jl:
```diff
@@ -10,12 +10,13 @@ end
+const MULTICOLUMNINDEX_TUPLE = (:AbstractVector, :Regex, :Not, :Between, :All, :Colon)
 
 const COLUMN_INDICATOR = "`Symbol`, string or integer"
 const COLUMNS_INDICATOR = "`:`, `All`, `Between`, `Not`, a regular expression," *
```
fixed

pdeffebach commented on docs/src/man/getting_started.md:
````diff
@@ -89,15 +106,28 @@ julia> df[:, firstcolumn] == df.A
+To get column names as `Symbol`s use the `propertynames` function:
+```
+julia> propertynames(df)
+(:A, :B)
````
```suggestion
2-element Array{Symbol,1}:
 :A
 :B
```

oxinabox commented:
Can a list of all the other things this PR changes be added to the top post?
Seems like it also reflows a bunch of docstrongs?
And I see some uses of `size` have been changed to `nrows`

bkamins commented on docs/src/man/getting_started.md:
```diff
@@ -58,6 +59,13 @@ julia> df.A
  3
  4
 
+julia> df."A"
```
I will mark it as resolved, as this is essentially the point of this PR to allow it (and I assume since @StefanKarpinski first proposed to accept `df."A"` it will keep getting supported by Julia parser :smile:).

Essentially the current recommendation was to write `df.var"some variable with spaces"` so we are just avoiding writing `var` in this case.

Indeed my initial fear was that allowing strings and `Symbol`s can lead to confusion but I think it is just a matter of documentation and tutorials.

bkamins commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -333,9 +342,17 @@ Return the number of dimensions of a data frame, which is always `2`.
 
+# separate methods are needed due to dispatch ambiguity
 Base.getproperty(df::AbstractDataFrame, col_ind::Symbol) = df[!, col_ind]
+Base.getproperty(df::AbstractDataFrame, col_ind::AbstractString) = df[!, col_ind]
```
If we allow strings for column indexing we should allow this also, so I resolve this (again `df."some col"` is a shorthand equivalent for for `df.var"some col"`).

bkamins commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -430,35 +447,37 @@ where each row represents a variable and each column a summary statistic.
-If the column does not allow missing values, `nothing` is returned.
-Consequently, `nmissing = 0` indicates that the column allows
-missing values, but does not currently contain any.
+Missing values are filtered in the calculation of all statistics, however the
```
no - it was on purpose in this PR. The docstrings were reformatted not to be overwide.

bkamins commented on src/dataframerow/dataframerow.jl:
```diff
@@ -154,26 +160,34 @@ function Base.haskey(r::DataFrameRow, key::Symbol)
+Base.haskey(r::DataFrameRow, key::AbstractString) = haskey(r, Symbol(key))
+
+# separate methods are needed due to dispatch ambiguity
+Base.getproperty(r::DataFrameRow, idx::Symbol) = r[idx]
```
Good point - I will att `SymbolOrString` union and use it.

bkamins replied:
Ah - now I remember why this is not possible for these three methods - there is method ambiguity problem. But in other places where we used `Union` I have added `SymbolOrString` as you have proposed.

bkamins commented:
@oxinabox - the problem we had was that there was never a good time to clean up the code and this PR anyway touched almost all files in the code-base (otherwise there was always a problem that such clean-up changes later force rebasing of all PRs, and with this PR we have stalled merging of other PRs anyway till this one is finished to make finishing it easier + I had to read every line in the code so when I spotted something I just corrected it).

E.g. I have changed `size` to `nrow` (I think it was only one place) - where `size` was mixed with `ncol` in the same expression. The change was from:
```
"<p>$(digitsep(size(df, 1))) rows × $(digitsep(ncol(df))) columns$omitmsg</p>"
```
to
```
"<p>$(digitsep(nrow(df))) rows × $(digitsep(ncol(df))) columns$omitmsg</p>"
```

So the list of changes in this PR is:

User visible:
* allow using an `AbstractString` **everywhere** where `Symbol` was accepted as a column indicator (i.e. this extends beyond just `getindex` and `getproperty`, every function was reviewed and made consistent)
* docstrings were updated to reflect this change + some unrelated docs fixes were added where spotted
* `names` function consistently returns vector of strings for all types that return column names
* `propertynames` function consistently returns vector of `Symbol`s for all types that return column names
* `keys` function consistently returns vector of `Symbol`s for all types that return column names as keys
* depend on DataAPI.jl v0.1.2
* `rename!` passes `String` to a function if API with function is used
* consistently define `hasproperty` for types that have custom `getproperty`
* add `cols` to `names` function for types where it makes sense consistently
* `push!` accepts dicts with string keys
* `indexing with an arbitrary eltype vector now throws better error messages (informtive `ArgumentError` is thrown in more cases rather than `MethodError` in the past)

Internal:
* add definitions of `SymbolOrString`, `ColumnIndex` (updated), and `MultiColumnIndex` unions and use them consistently across whole codebase
* remove `keys` from `AbstractIndex` as it was never used
* correct formatting of code in many places (mainly docstrings, function signatures, and adding missing `return` statements in long functions as many of them were updated anyway, but also in unrelated places)
* fix a rule that `_names` returns vector of `Symbol`s without copying
* in some places switch from `names` to `_names` internally to avoid unnecessary allocations
* in some places use `groupcols` where this function would be an appropriate to call; similarly with `names(df, cols)` where it made sense
* remove `without` function that was not needed anymore

(I will make this list in the top post also)

bkamins commented:
> I'm not really convinced

I am largely indifferent here but since in base we have `VecOrMat` I think it is OK, and @pdeffebach found it more readable (and for sure it is shorter so we do not have to break lines so oftern in the code).

bkamins commented:
Thank you all for working on this huge PR!

oxinabox commented on src/dataframerow/dataframerow.jl:
```diff
@@ -111,15 +111,18 @@ for T in (:AbstractVector, :Regex, :Not, :Between, :All, :Colon)
         if v isa AbstractDict
+            if all(x -> x isa AbstractString, keys(v))
+                v = (;(Symbol.(keys(v)) .=> values(v))...)
+            end
             for n in view(_names(df), idxs)
                 if !haskey(v, n)
                     throw(ArgumentError("Column :$n not found in source dictionary"))
                 end
             end
         elseif !all(((a, b),) -> a == b, zip(view(_names(df), idxs), keys(v)))
             mismatched = findall(view(_names(df), idxs) .!= collect(keys(v)))
-            throw(ArgumentError("Selected column names do not match the names in assigned value in" *
-                                " positions $(join(mismatched, ", ", " and "))"))
+            throw(ArgumentError("Selected column names do not match the names in assigned " *
+                                "value in positions $(join(mismatched, ", ", " and "))"))
         end
 
         for (col, val) in pairs(v)
```
An alterantive way to do this would be to declare a local variable
`v_keys` which promises to be a collection of `Symbol`.
e.g. 
```
v_keys = keys(v)
v_keys = keytype(v) === Symbol ? keys(v) : Symbol.(v)
```

then each time `keys(v)` is used in this function it can be replaced with `v_keys`
The values are never used until the end

This would save having to convert the `values` as well,
and line 128 works already with symbols or strings

bkamins replied:
this is related to the comment above - if we allow mixing strings with `Symbol`s.
Currently we do not allow mixing. When we decide on this I will rewrite it anyway.

bkamins commented:
Thank you for the comments. I will make a separate PR taking them into account.

bkamins commented on docs/src/lib/indexing.md:
```diff
@@ -230,3 +232,18 @@ The elements of a `GroupedDataFrame` are [`SubDataFrame`](@ref)s of its parent.
+|---------------------|------------------|------------------|------------------|
+| `AbstractDataFrame` | `Vector{String}` | `Vector{Symbol}` | undefined        |
+| `DataFrameRow`      | `Vector{String}` | `Vector{Symbol}` | `Vector{Symbol}` |
+| `DataFrameRows`     | `Vector{String}` | `Vector{Symbol}` | vector of `Int`  |
```
Actually it is `LinearIndices{1,Tuple{Base.OneTo{Int}}}` but I wanted to avoid writing this as it would be confusing I think.

bkamins commented on docs/src/lib/indexing.md:
```diff
@@ -230,3 +232,18 @@ The elements of a `GroupedDataFrame` are [`SubDataFrame`](@ref)s of its parent.
+| `DataFrameRows`     | `Vector{String}` | `Vector{Symbol}` | vector of `Int`  |
+| `DataFrameColumns`  | `Vector{String}` | `Vector{Symbol}` | `Vector{Symbol}` |
+| `GroupedDataFrame`  | `Vector{String}` | tuple of fields  | `GroupKeys`      |
+| `GroupKeys`         | undefined        | tuple of fields  | vector of `Int`  |
```
same here - it is `LinearIndices{1,Tuple{Base.OneTo{Int}}}`

bkamins commented on docs/src/man/getting_started.md:
````diff
@@ -45,7 +45,8 @@ julia> df = DataFrame(A = 1:4, B = ["M", "F", "F", "M"])
 ```
 
-Columns can be directly (i.e. without copying) accessed via `df.col` or `df[!, :col]`. The latter syntax is more flexible as it allows passing a variable holding the name of the column, and not only a literal name. Note that column names are symbols (`:col` or `Symbol("col")`) rather than strings (`"col"`). Columns can also be accessed using an integer index specifying their position.
+Columns can be directly (i.e. without copying) accessed via `df.col`, `df."col"`, `df[!, :col]` or `df[!, "col"]`. The two latter syntaxes are more flexible as they allow passing a variable holding the name of the column, and not only a literal name. Note that column names can be either symbols (written as `:col`, `:var"col"` or `Symbol("col")`) or strings (written as `"col"`).
````
It does not work. I will add a note in the PR that follows your comments.

bkamins commented on docs/src/man/getting_started.md:
```diff
@@ -89,15 +106,30 @@ julia> df[:, firstcolumn] == df.A
+
+    DataFrames.jl allows to use `Symbol`s (like `:A`) and strings (like `"A"`)
+    for all column indexing operations for convenience.
+    However, using `Symbol`s is slightly faster and should generally be preferred.
```
added

bkamins commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -538,22 +559,23 @@ function _describe(df::AbstractDataFrame, stats::AbstractVector)
     # Put the summary stats into the return data frame
     data = DataFrame()
-    data.variable = names(df)
+    data.variable = copy(_names(df))
```
good point - fixed.

bkamins commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -1330,16 +1368,20 @@ julia> vcat(d4, df1)
 """
 Base.vcat(dfs::AbstractDataFrame...;
-          cols::Union{Symbol, AbstractVector{Symbol}}=:setequal) =
+          cols::Union{Symbol, AbstractVector{Symbol},
+                      AbstractVector{<:AbstractString}}=:setequal) =
```
This will allow mixing strings and `Symbol`s. I leave it out for now as I want to hear what @nalimilan thinks about it (I will make a note in the PR to consider this).
